godot-macros/src/docs/md2bb.rs

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! Converts [markdown](https://en.wikipedia.org/wiki/Markdown) to [BBCode](https://en.wikipedia.org/wiki/BBCode).
+use markdown::mdast::Node;
+use markdown::{to_mdast, ParseOptions};
+use std::collections::HashMap;
+
+pub fn convert(md: &str) -> String {
+    let n = to_mdast(md, &ParseOptions::gfm()).unwrap();
+    let definitions = n
+        .children()
+        .unwrap()
+        .iter()
+        .filter_map(|n| match n {
+            Node::Definition(definition) => {
+                Some((definition.identifier.clone(), definition.url.clone()))
+            }
+            _ => None,
+        })
+        .collect::<HashMap<String, String>>();
+
+    n.children()
+        .unwrap()
+        .iter()
+        .filter_map(|n| node(n, &definitions))
+        .reduce(|a, b| format!("{a}\n{b}"))
+        .unwrap_or_default()
+}
+
+// will change to &mut String if necessary
+fn node(node: &Node, definitions: &HashMap<String, String>) -> Option<String> {
+    use Node::*;
+    Some(match node {
+        Root(root) => nodes(&root.children, definitions, ""),
+        BlockQuote(quote) => format!("“{}”", nodes(&quote.children, definitions, "\n\n")),
+        List(list) => {
+            let text = nodes(&list.children, definitions, "\n");
+            match list.ordered {
+                true => format!("[ol]\n{text}\n[/ol]"),
+                false => format!("[ul]\n{text}\n[/ul]"),
+            }
+        }
+        Break(_) => "\n".into(),
+        InlineCode(markdown::mdast::InlineCode { value, .. }) => format!("[code]{value}[/code]"),
+        Delete(delete) => format!("[s]{}[/s]", nodes(&delete.children, definitions, "")),
+        Emphasis(emphasis) => format!("[i]{}[/i]", nodes(&emphasis.children, definitions, "")),
+        Image(markdown::mdast::Image { url, .. }) => format!("[img]{url}[/img]",),
+        ImageReference(image) => {
+            format!("[img]{}[/img]", definitions.get(&image.identifier).unwrap())
+        }
+        Link(markdown::mdast::Link { url, children, .. }) => {
+            format!("[url={url}]{}[/url]", nodes(children, definitions, ""))
+        }
+        LinkReference(markdown::mdast::LinkReference {
+            identifier,
+            children,
+            ..
+        }) => format!(
+            "[url={}]{}[/url]",
+            definitions.get(identifier).unwrap(),
+            nodes(children, definitions, "")
+        ),
+        Strong(strong) => format!("[b]{}[/b]", nodes(&strong.children, definitions, "")),
+        Text(text) => text.value.clone(),
+        Code(markdown::mdast::Code { value, .. }) => format!("[codeblock]\n{value}\n[/codeblock]"),
```

Comment:
i think this will format as gdscript by default, we should probably check and only output this as appropriate when the user writes \`\`\`gdscript. or we can leave this be as-is and just add a `// TODO` here about that.

Reply:
It will only format as gdscript, however theres nothing you can do about that.

I think its nicer for it to be formatted in gdscript in a codeblock than not at all.

Reply:
you _can_ format as c# as well ig. maybe this would be useful to suggest as a proposal to godot? adding more options for codeblocks.